dashboard/lib/services/lesson_plan_pdfs.rb

```diff
@@ -0,0 +1,160 @@
+        generate_lesson_pdf(lesson, pdf_dir)
+      end
+
+      # TODO: Script Overview PDFs
+      #
+      # There are still some outstanding questions as to what exactly we
+      # want to do here. We know we want to provide a single PDF containing
+      # all lesson data (so users who want to print all lessons for a script
+      # can do so in a single print job), but it sounds like we also want
+      # some kind of script overview contenet and possibly also a "visual
+      # map" of the lesson plans.
+      #
+      # Notably, we haven't yet moved over course and script overview
+      # content from CurriculumBuilder nor built views for displaying that
+      # content. Once we do, we will be better prepared to figure out what
+      # kind of PDF to construct here.
+      upload_generated_pdfs_to_s3(pdf_dir)
+
+      FileUtils.remove_entry_secure(pdf_dir) unless DEBUG
+    end
+
+    def self.get_base_url
+      # Right now, this is obviously just using the raw S3 subdomain as the
+      # base url. This should work fine for now, but ideally we'd like to set
+      # the bucket up to be served from a code.org subdomain so we can have a
+      # button which downloads the asset (the HTML download attribute only
+      # works with same-origin urls).
+      "https://#{S3_BUCKET}.s3.amazonaws.com"
+    end
+
+    # Build the full path of the lesson plan PDF for the given lesson. This
+    # will be based on not only the lesson's script but also the current
+    # version of the script in the environment.
+    #
+    # Expect this to look something like
+    # <Pathname:csp1-2021/20210216001309/Welcome to CSP.pdf>
+    def self.get_pathname(lesson)
+      return nil unless lesson&.script&.seeded_at
+      version_number = Time.parse(lesson.script.seeded_at).to_s(:number)
+      filename = ActiveStorage::Filename.new(lesson.key + ".pdf").sanitized
+      return Pathname.new(File.join(lesson.script.name, version_number, filename))
+    end
+
+    def self.get_url(lesson)
+      pathname = get_pathname(lesson)
+      return nil unless pathname.present?
+
+      File.join(get_base_url, pathname)
+    end
+
+    def self.generate_lesson_pdf(lesson, directory="/tmp/")
+      url = Rails.application.routes.url_helpers.script_lesson_url(lesson.script, lesson)
+      pathname = get_pathname(lesson)
+
+      ChatClient.log "Generating #{pathname.to_s.inspect} from #{url.inspect}"
+
+      FileUtils.mkdir_p(File.join(directory, pathname.dirname))
+      PDF.generate_from_url(url, File.join(directory, pathname))
+    end
+
+    def self.upload_generated_pdfs_to_s3(directory)
+      ChatClient.log "Uploading all generated PDFs to S3"
+      ChatClient.log "from local temporary directory #{directory.inspect}" if DEBUG
+      Dir.glob(File.join(directory, '**/*.pdf')).each do |filepath|
+        # Note that this "filename" includes subdirectories; this is fine
+        # only because we're uploading to S3.
```

Comment:
I don't understand this comment. I doubt you need to change it but for my own education, why would it be not-fine if we weren't uploading to S3?

Reply:
We're generating a filename that looks something like `"csd1-2021/12309239123123/Lesson Name.pdf"`. Note, however, that that string is more like a full file _path_ than a simple filename. Ordinarily we'd have to make sure that we were explicitly accomodating the directory structure, but this is actually how S3 does things. S3 doesn't actually have a directory structure at all, it's an entirely flat file storage system. But what it does do is _emulate_ a directory structure by treating filenames as though they were full file paths.

The comment is meant to call out that we're doing something a little weird here, but it's okay because it's the exact kind of weirdness that S3 is okay with us doing.